app/src/main/java/com/asfoundation/wallet/ui/gamification/GamificationPresenter.kt

```diff
@@ -47,79 +48,92 @@ class GamificationPresenter(private val view: GamificationView,
-                  currentLevel: Int): Pair<List<LevelModel>, List<LevelModel>> {
-    val hiddenList = ArrayList<LevelModel>()
-    val shownList = ArrayList<LevelModel>()
-    for (level in levels) {
-      val viewType = when {
-        level.level < currentLevel -> LevelType.REACHED
-        level.level == currentLevel -> LevelType.CURRENT
-        else -> LevelType.UNREACHED
+  private fun displayHeaderInformation(gamification: GamificationInfo) {
+    if (gamification.totalEarned != null) {
+      val totalSpent = formatter.formatCurrency(gamification.totalSpend, WalletCurrency.FIAT)
+      val bonusEarned =
+          formatter.formatCurrency(gamification.totalEarned.amount, WalletCurrency.FIAT)
+      view.showHeaderInformation(totalSpent, bonusEarned, gamification.totalEarned.symbol)
+    }
+  }
+
+  private fun mapToLevelsList(
+      gamification: GamificationInfo): Pair<List<LevelItem>, List<LevelItem>> {
+    val hiddenList = ArrayList<LevelItem>()
+    val shownList = ArrayList<LevelItem>()
+    val currentLevel = gamification.currentLevel
+    for (level in gamification.levels) {
+      val levelItem = when {
+        level.level < currentLevel -> ReachedLevelItem(level.amount, level.bonus, level.level)
+        level.level == currentLevel -> CurrentLevelItem(level.amount, level.bonus, level.level,
+            gamification.totalSpend, gamification.nextLevelAmount)
+        else -> UnreachedLevelItem(level.amount, level.bonus, level.level)
       }
-      val levelViewModel = LevelModel(level.amount, level.bonus, level.level, viewType)
-      if (viewType == LevelType.REACHED) hiddenList.add(levelViewModel)
-      else shownList.add(levelViewModel)
+      if (levelItem is ReachedLevelItem) hiddenList.add(levelItem)
+      else shownList.add(levelItem)
     }
     return Pair(hiddenList, shownList)
   }
 
-  private fun handleHeaderInformation(totalEarned: BigDecimal, totalSpend: BigDecimal,
-                                      status: GamificationStats.Status) {
-    if (status == GamificationStats.Status.OK) {
-      disposables.add(gamification.getAppcToLocalFiat(totalEarned.toString(), 2)
-          .filter { it.amount.toInt() >= 0 }
-          .observeOn(viewScheduler)
-          .doOnNext {
-            val totalSpent = formatter.formatCurrency(totalSpend, WalletCurrency.FIAT)
-            val bonusEarned = formatter.formatCurrency(it.amount, WalletCurrency.FIAT)
-            view.showHeaderInformation(totalSpent, bonusEarned, it.symbol)
+  private fun getUserStatsAndBonusEarned(): Observable<Pair<GamificationStats, FiatValue>> {
+    return gamification.getUserStats()
+        .flatMap { stats ->
+          if (stats.status == GamificationStats.Status.OK) {
+            gamification.getAppcToLocalFiat(stats.totalEarned.toString(), 2, stats.fromCache)
+                .map { Pair(stats, it) }
+          } else {
+            Observable.just(Pair(stats, FiatValue(BigDecimal.ONE.negate(), "")))
```

Comment:
I'm gonna leave this up to you, as we discussed we could refactor this so that fiatValue is is instead inside GamificationStats and is handled by its repository. This way we don't need this complicated scheme of caching the conversion rate and also avoid having the conversion rate cache be handled by whoever calls it.